lib/resty/mlcache.lua

```diff
@@ -222,9 +222,51 @@ function _M.new(name, shm, opts)
         end
 
         if opts.ipc_shm ~= nil and type(opts.ipc_shm) ~= "string" then
+            -- Support for deprecated `ipc_shm` option.
+            -- TODO: eventually remove.
             error("opts.ipc_shm must be a string", 2)
         end
 
+        if opts.ipc ~= nil then
+            if type(opts.ipc) ~= "table" then
+                error("opts.ipc must be a table", 2)
+            end
+
+            if opts.ipc.type ~= "mlcache_ipc"
+                and opts.ipc.type ~= "custom"
+            then
+                error("opts.ipc.type must be one of 'mlcache_ipc' or 'custom'",
+                      2)
+            end
+
+            if opts.ipc.type == "mlcache_ipc"
+                and type(opts.ipc.shm) ~= "string"
+            then
+                error("opts.ipc.shm must be a string", 2)
+
+            elseif opts.ipc.type == "custom" then
+                if type(opts.ipc.register_listeners) ~= "function" then
+                    error("opts.ipc.register_listeners must be a function", 2)
+                end
+
+                if type(opts.ipc.broadcast) ~= "function" then
+                    error("opts.ipc.broadcast must be a function", 2)
+                end
+
+                if type(opts.ipc.poll) ~= "function" then
+                    error("opts.ipc.poll must be a function", 2)
+                end
+            end
```

Comment:
in the above the user sets "mlcache_ipc" or "custom". Wouldn't it be easier to have a switch like "enable_ipc", and then either 3 functions are provided, or if not it defaults to "mlcache_ipc"?

Reply:
In the future, if and when we support more IPC mechanisms (ngx_lua_ipc or even maybe a native OpenResty one once cosockets can bind), we might want to specify arbitrary options to those modules, hence why `opts.ipc` is a loose table. Even right now, the mlcache_ipc mode has:
```lua
opts.ipc = {
  type = "mlcache_ipc",
  shm = "ipc_shm",
  debug = true
}
```

So this way we can namespace multiple IPC configuration values into one table.

Reply:
I'm not sure I follow. The table is loose, but the api is still 3 functions in that table; `register_listeners`, `poll`, and `broadcast`. Those same functions can be created for `ngx_lua_ipc` or any other right? so you can provide any arbitrary option you like. Naming specific implementations (and possibly implementing specifics) becomes a maintenance burden in my experience, hence my comment.